File: exercism/zig/armstrong-numbers/armstrong_numbers.zig
const std = @import("std");

pub fn isArmstrongNumber(num: u128) bool {
    if (num == 0) return true;

    // log2(log10(2^128)) is ~5.3
    const numf64: f64 = @floatFromInt(num);
    const digits: u6 = @intFromFloat(@ceil(@log10(numf64)));
    var sum: u129 = 0;
    var n = num;
    while (n != 0) : (n /= 10) {
        sum += std.math.pow(u128, n % 10, digits);
    }
    std.debug.print("\nsum={d}\n", .{sum});
    return sum == num;
}

File: exercism/c/armstrong-numbers/armstrong_numbers.c
#include "armstrong_numbers.h"
#include <math.h>

bool is_armstrong_number(int candidate)
{
   int digits = log10(candidate) + 1;

   int sum = 0;
   for (int i = candidate; i > 0; i /= 10)
   {
      sum += power(i % 10, digits);
   }
   return sum == candidate;
}

int power(int base, int exponent) {
   if (exponent == 0) {
      return 1;
   }
   if (exponent % 2 == 1) {
      return base * power(base * base, (exponent - 1) / 2);
   } else {
      return power(base * base, exponent / 2);
   }
}
